tools/language-server/wstest/src/main.rs

```diff
@@ -0,0 +1,311 @@
+mod format;
+
+use std::path::PathBuf;
+use std::time::Duration;
+
+use clap::Parser;
+use clap::ValueHint;
+use futures::SinkExt;
+use futures::StreamExt;
+use tokio::io::AsyncBufReadExt;
+use tokio::io::BufReader;
+use tokio::sync::mpsc;
+use url::Url;
+use websocket_lite::ClientBuilder;
+use websocket_lite::Message;
+use websocket_lite::Opcode;
+use websocket_lite::Result;
+
+static EXPECT_TEXT_RESPONSE: &str = "t|";
+static EXPECT_BINARY_RESPONSE: &str = "b|";
+
+#[derive(Parser, Debug)]
+#[clap(version, about)]
+struct Args {
+    /// Text websocket to connect to
+    #[clap(value_name = "URL", value_hint = ValueHint::Url)]
+    text_socket: Url,
+
+    /// File containing messages to initialize the main socket
+    #[clap(long, value_hint = ValueHint::FilePath)]
+    init_text_socket: Option<PathBuf>,
+
+    /// File containing responses to ignore
+    #[clap(long, value_hint = ValueHint::FilePath)]
+    ignore_text_socket_responses: Option<PathBuf>,
+
+    /// Binary socket to connect to
+    #[clap(long, value_name = "URL", value_hint = ValueHint::Url)]
+    binary_socket: Option<Url>,
+
+    /// File containing messsages to initialize the binary socket
+    #[clap(long, value_hint = ValueHint::FilePath)]
+    init_binary_socket: Option<PathBuf>,
+
+    /// Path to a file with commands to send
+    #[clap(long, value_name = "FILE", value_hint = ValueHint::FilePath)]
+    input: Option<PathBuf>,
+
+    /// Input commands expect responses from a binary socket
+    #[clap(long)]
+    input_expects_binary_responses: bool,
+
+    /// Number of warmup requests to send
+    #[clap(long, default_value = "5")]
+    warmup_iterations: usize,
+
+    /// Number of benchmarked requests to send
+    #[clap(long, default_value = "5")]
+    benchmark_iterations: usize,
+
+    /// Time in milliseconds to wait before sending input messages (after the init sequence is
+    /// complete)
+    #[clap(long, value_name = "MILLISECONDS", default_value = "0")]
+    wait_after_init: u64,
+
+    /// Time in milliseconds to wait before starting benchmark (after the warmup is complete)
+    #[clap(long, value_name = "MILLISECONDS", default_value = "0")]
+    wait_after_warmup: u64,
+
+    /// Time in milliseconds to wait before sending the next request from the `input` file
+    #[clap(long, value_name = "MILLISECONDS", default_value = "0")]
+    wait_after_response: u64,
+}
+
+/// Synchronization messages used for establishing order between requests and responses
+#[derive(Debug)]
+enum SyncMessage {
+    ResponseReceived,
+}
+
+/// Read file line by line
+async fn read_lines(path_buf: PathBuf) -> Result<Vec<String>> {
+    let file = tokio::fs::File::open(path_buf.as_path()).await?;
+    let mut lines = BufReader::new(file).lines();
+
+    let mut res = vec![];
+    while let Some(line) = lines.next_line().await? {
+        res.push(line);
+    }
```

Comment:
Another place for Stream and StreamExt. Please go throug remaining loops and change them to Stream or Iterator operations where plausible.

Reply:
Ok, we made that work in the end